quaternions/quaternions.py

```diff
@@ -293,21 +293,75 @@ def from_qmethod(source, target, probabilities=None):
 
         See Averaging Quaternions, by Markley, Cheng, Crassidis, Oschman.
         '''
         B = np.array([q.coordinates for q in quaternions])
-        if not weights:
+        if not any(True for _ in weights):
             weights = np.ones(len(quaternions))
         M = B.T.dot(np.diag(weights)).dot(B)
 
         return Quaternion._first_eigenvector(M)
 
     @staticmethod
+    def average_and_std_naive(*quaternions, weights=()):
+        """
+        Naive implementation of std. dev. calculation, assuming small angles between quaternions
+        and average. Returns average quaternion and stddev of input quaternion list in deg
+        """
+        if not any(True for _ in weights):
+            weights = np.ones(len(quaternions))
+        q_average = Quaternion.average(*quaternions, weights=weights)
+        diffs = np.array([q_average.distance(quat) for quat in quaternions])
+        # TODO: check that this way of taking into account weights in stddev calculation is ok
+        stddev = np.degrees(np.sqrt(sum((diffs ** 2) * weights) / np.sum(weights)))
+        return q_average, stddev
+
+    @staticmethod
+    def average_and_std_lerner(*quaternions, weights=()):
+        """
+        Calculates the std. dev. as 1.87 of the root of the maximum eigenvalue of the covariance matrix,
+        called sigma lerner. This covariance matriz is the one asociated with the rotation angles of
+        each input quaternion agains the average one. Returns the average quaternion anf the sigma lerner
+        in deg.
+        :param quaternions: lista de objetos Quaternion
+        :param weights:
+        :return: average quaternions, sigma lerner
+        """
+        q_average = Quaternion.average(*quaternions, weights=weights)
+        diffs = [quat / q_average for quat in quaternions]
+        angles_list = np.array([[2 * diff.coordinates[i] for i in [1, 2, 3]] for diff in diffs])
+        covariance_matrix = covariance_matrix_from_angles(angles_list)
+        return q_average, np.degrees(sigma_lerner(covariance_matrix))
+
+    @staticmethod
+    def average_and_covariance(*quaternions, R=np.eye(3)):
+        '''
+        Returns the quaternion such that its matrix minimizes the square distance
+        to the matrices of the quaternions in the argument list, and the resulting
+        covariance matrix asociated with that quaternion. Input matrix R (3x3) is
+        assume to be the same for all input quaternions, and in rads**2
+
+        See Averaging Quaternions, by Markley, Cheng, Crassidis, Oschman.
+        '''
+        q_average = Quaternion.average(*quaternions)
+
+        R_inverse = np.linalg.inv(R)
+        orthogonal_matrix_sum = np.zeros((4, 4))
+        for q in quaternions:
+            orthogonal_matrix_sum += orthogonal_matrix(q).dot(
+                R_inverse.dot(orthogonal_matrix(q).T))
+
+        cov_dev_matrix = np.linalg.inv(orthogonal_matrix(q_average).T.dot(
```

Comment:
I guess the code is required to run in Python 3.4? Otherwise you could use `orthogonal_matrix(q_average).T @ orthogonal_matrix_sum @ orthogonal_matrix(q_average)`.

Reply:
yeap, I still fun python 3.4 in my debian8 installation :flushed: Do we have a requisite on the python version?